cdap-data-fabric/src/test/java/co/cask/cdap/spi/metadata/MetadataStorageTest.java

```diff
@@ -0,0 +1,1101 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spi.metadata;
+
+import co.cask.cdap.api.metadata.MetadataEntity;
+import co.cask.cdap.api.metadata.MetadataScope;
+import co.cask.cdap.proto.element.EntityTypeSimpleName;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.spi.metadata.MetadataMutation.Drop;
+import co.cask.cdap.spi.metadata.MetadataMutation.Remove;
+import co.cask.cdap.spi.metadata.MetadataMutation.Update;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static co.cask.cdap.api.metadata.MetadataScope.SYSTEM;
+import static co.cask.cdap.api.metadata.MetadataScope.USER;
+import static co.cask.cdap.data2.metadata.system.SystemMetadataProvider.ENTITY_NAME_KEY;
+import static co.cask.cdap.spi.metadata.MetadataKind.PROPERTY;
+import static co.cask.cdap.spi.metadata.MetadataKind.TAG;
+
+/**
+ * Tests for Metadata SPI implementations.
+ */
+public abstract class MetadataStorageTest {
+
+  private static final String TYPE_ARTIFACT = EntityTypeSimpleName.ARTIFACT.name();
+  private static final String TYPE_DATASET = EntityTypeSimpleName.DATASET.name();
+  private static final String TYPE_PROGRAM = EntityTypeSimpleName.PROGRAM.name();
+
+  protected abstract MetadataStorage getMetadataStorage();
+
+  @After
+  public void ensureCleanUp() throws IOException {
+    assertEmpty(getMetadataStorage(), SearchRequest.of("*").build());
+  }
+
+  @Test
+  public void testMutations() throws IOException {
+    MetadataStorage mds = getMetadataStorage();
+
+    MetadataEntity entity = NamespaceId.DEFAULT.dataset("entity").toMetadataEntity();
```

Comment:
hmmm. I was thinking about that but: In CDAP the typical usage will be with MetadataEntities that are converted from EntityIds. That is why I decided to keep it that way. Do you think it changes the way the metadata store works?

Reply:
I will open a Jira to add more tests for non-CDAP metadata entities. I think we will need them anyway, because we need to test that searching for all entity types returns custom entities (that do not correspond to any of the SimpleEntityName enum).

Reply:
> In CDAP the typical usage will be with MetadataEntities that are converted from EntityIds. 

True. Let's keep it then. It does not changes how metadata store work.